src/Lecture2.hs

```diff
@@ -175,64 +315,97 @@ challenging. However, you still may find some of them easier than some of the
+--
+-- The lists are guaranteed to be given sorted, so you don't need to
+-- verify that.
+--
+-- >>> merge [1, 2, 4] [3, 7]
+-- [1,2,3,4,7]
 merge :: [Int] -> [Int] -> [Int]
-merge = error "TODO"
-
-{- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
-function takes a list of numbers and returns a new list containing the
-same numbers but in the increasing order.
-
-The algorithm of merge sort is the following:
-
-  1. If the given list has less than 2 elements, it's already sorted.
-  2. Otherwise, split list into two lists of the same size.
-  3. Sort each of two lists recursively.
-  4. Merge two resulting sorted lists to get a new sorted list.
-
->>> mergeSort [3, 1, 2]
-[1,2,3]
--}
+merge xs [] = xs
+merge [] ys = ys
+merge (x : xs) (y : ys)
+  | x == y = x : y : merge xs ys
+  | x < y = x : merge xs (y : ys)
+  | otherwise = y : merge (x : xs) ys
+
+-- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
+-- function takes a list of numbers and returns a new list containing the
+-- same numbers but in the increasing order.
+--
+-- The algorithm of merge sort is the following:
+--
+--  1. If the given list has less than 2 elements, it's already sorted.
+--  2. Otherwise, split list into two lists of the same size.
+--  3. Sort each of two lists recursively.
+--  4. Merge two resulting sorted lists to get a new sorted list.
+--
+-- >>> mergeSort [3, 1, 2]
+-- [1,2,3]
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
-
-
-{- | Haskell is famous for being a superb language for implementing
-compilers and interpreters to other programming languages. In the next
-tasks, you need to implement a tiny part of a compiler.
-
-We're going to work on a small subset of arithmetic operations.
-
-In programming we write expressions like "x + 1" or "y + x + 10".
-Such expressions can be represented in a more structured way (than a
-string) using the following recursive Algebraic Data Type:
--}
+mergeSort = mergeSortSingletons
+
+-- mergeSort = mergeSortEvenOdd
+
+-- Solution 1 : Instead of using length to split the list in the middle,
+-- we can split by index.
+splitByEvenOddIndex :: [a] -> ([a], [a])
+splitByEvenOddIndex [] = ([], [])
+splitByEvenOddIndex [x] = ([x], [])
+splitByEvenOddIndex (x : y : zs) = (x : xs, y : ys)
+  where
+    (xs, ys) = splitByEvenOddIndex zs
+
+mergeSortEvenOdd :: [Int] -> [Int]
+mergeSortEvenOdd [] = []
+mergeSortEvenOdd [x] = [x]
+mergeSortEvenOdd l = merge (mergeSortEvenOdd evenl) (mergeSortEvenOdd oddl)
+  where
+    (evenl, oddl) = splitByEvenOddIndex l
+
+-- Solution 2 : Alternatively, we can split every element in a singleton list
+-- and merge 2 by 2
+mergeSortSingletons :: [Int] -> [Int]
+mergeSortSingletons [] = []
+mergeSortSingletons [x] = [x]
+mergeSortSingletons l = mergeSortConcat (map singleton l)
+  where
+    mergeSortConcat :: [[Int]] -> [Int]
+    mergeSortConcat [] = []
+    mergeSortConcat [x] = x
+    mergeSortConcat (x : y : zs) = mergeSortConcat (merge x y : zs)
```

Comment:
This implementation actually won't have `O(n log n)` complexity and will be `O(n^2)` because you always merge the first two lists.

Compare:

```haskell
[1, -1, 2, -2]
=> [[1], [-1], [2], [-2]]
=> [[-1, 1], [2], [-2]]
=> [[-1, 1, 2 ], [-2]]
=> [[-2, -1, 1, 2 ]]
```

So it's more like an _insertion sort_ where you just merge your head with a singleton list on every iteration.

While you actually should merge _every two pairs_, like this:

```haskell
[1, -1, 2, -2]
=> [[1], [-1], [2], [-2]]
=> [[-1, 1], [-2, 2]]
=> [[-2, -1, 1, 2 ]]
```